File: algorithms/algorithms-cplusplus/leetcode/Question_0282_Expression_Add_Operators.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<string> addOperators(string num, int target) {
    vector<string> res;
    if (!num.empty()) {
      vector<string> ops;
      help(num, target, 0, "", 0, 0, res);
    }
    return res;
  }

 private:
  void help(string& num, int target, int i, string path, long cur, long pre, vector<string>& res) {
    if (i == num.size()) {
      if (cur == target) {
        res.push_back(path);
      }
      return;
    }
    long v = 0;
    string vStr;
    for (int j = i; j < num.length(); j++) {
      if (j > i && num[i] == '0') {
        break;
      }
      v = v * 10 + (num[j] - '0');
      vStr += num[j];
      if (i == 0) {
        help(num, target, j + 1, path + vStr, v, v, res);
      } else {
        help(num, target, j + 1, path + "+" + vStr, cur + v, v, res);
        help(num, target, j + 1, path + "-" + vStr, cur - v, -v, res);
        help(num, target, j + 1, path + "*" + vStr, cur - pre + pre * v, pre * v, res);
      }
    }
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0282_Expression_Add_Operators.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0282_Expression_Add_Operators {
    public List<String> addOperators(String num, int target) {
        List<String> res = new ArrayList<>();
        if (num.length() == 0) {
            return res;
        }
        help(num, target, 0, 0, 0, 0, new ArrayList<>(), res);
        return res;
    }

    private void help(String num, int target, int i, long pre, long cur, long v, List<String> ops, List<String> res) {
        if (i == num.length()) {
            if (v == target && cur == 0) {
                StringBuilder sb = new StringBuilder();
                ops.subList(1, ops.size()).forEach(s -> sb.append(s));
                res.add(sb.toString());
            }
            return;
        }

        cur = cur * 10 + Character.getNumericValue(num.charAt(i));
        String curStr = String.valueOf(cur);
        if (cur > 0) {
            help(num, target, i + 1, pre, cur, v, ops, res);
        }

        ops.add("+");
        ops.add(curStr);
        help(num, target, i + 1, cur, 0, v + cur, ops, res);
        ops.remove(ops.size() - 1);
        ops.remove(ops.size() - 1);

        if (ops.size() > 0) {
            ops.add("-");
            ops.add(curStr);
            help(num, target, i + 1, -cur, 0, v - cur, ops, res);
            ops.remove(ops.size() - 1);
            ops.remove(ops.size() - 1);

            ops.add("*");
            ops.add(curStr);
            help(num, target, i + 1, cur * pre, 0, v - pre + (cur * pre), ops, res);
            ops.remove(ops.size() - 1);
            ops.remove(ops.size() - 1);
        }
    }
}
